File: backend/src/main/java/com/rodrigocoelhoo/lifemanager/training/service/TrainingPlanService.java
package com.rodrigocoelhoo.lifemanager.training.service;

import com.rodrigocoelhoo.lifemanager.config.RedisCacheService;
import com.rodrigocoelhoo.lifemanager.exceptions.ResourceNotFound;
import com.rodrigocoelhoo.lifemanager.training.dto.trainingplandto.TrainingPlanDTO;
import com.rodrigocoelhoo.lifemanager.training.dto.trainingplandto.TrainingPlanResponseDTO;
import com.rodrigocoelhoo.lifemanager.training.dto.trainingplandto.TrainingPlanUpdateDTO;
import com.rodrigocoelhoo.lifemanager.training.model.ExerciseModel;
import com.rodrigocoelhoo.lifemanager.training.model.TrainingPlanModel;
import com.rodrigocoelhoo.lifemanager.training.repository.TrainingPlanRepository;
import com.rodrigocoelhoo.lifemanager.users.UserModel;
import com.rodrigocoelhoo.lifemanager.users.UserService;
import jakarta.transaction.Transactional;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.List;

@Service
public class TrainingPlanService {

    private final TrainingPlanRepository trainingPlanRepository;
    private final UserService userService;
    private final ExerciseService exerciseService;
    private final RedisCacheService redisCacheService;

    private static final String CACHE_LIST = "trainingPlans";

    public TrainingPlanService(
            TrainingPlanRepository trainingPlanRepository,
            UserService userService,
            ExerciseService exerciseService,
            RedisCacheService redisCacheService
    ) {
        this.trainingPlanRepository = trainingPlanRepository;
        this.userService = userService;
        this.exerciseService = exerciseService;
        this.redisCacheService = redisCacheService;
    }

    @Cacheable(value = CACHE_LIST, keyGenerator = "userAwareKeyGenerator")
    public Page<TrainingPlanResponseDTO> getAllTrainingPlansByUser(Pageable pageable) {
        UserModel user = userService.getLoggedInUser();
        Page<TrainingPlanModel> page = trainingPlanRepository.findAllByUser(user, pageable);
        return page.map(TrainingPlanResponseDTO::fromEntity);
    }

    public TrainingPlanModel getTrainingPlan(Long id) {
        UserModel user = userService.getLoggedInUser();
        return trainingPlanRepository.findByIdAndUser(id, user)
                .orElseThrow(() -> new ResourceNotFound("Training plan with ID '" + id + "' does not belong to the current user"));
    }

    @Transactional
    public TrainingPlanModel createTrainingPlan(TrainingPlanDTO data) {
        UserModel user = userService.getLoggedInUser();
        TrainingPlanModel trainingPlanModel = TrainingPlanModel.builder()
                .user(user)
                .name(data.name())
                .description(data.description())
                .exercises(List.of())
                .build();

        TrainingPlanModel saved = trainingPlanRepository.save(trainingPlanModel);

        redisCacheService.evictUserCache(CACHE_LIST);

        return saved;
    }

    @Transactional
    public TrainingPlanModel updateTrainingPlan(Long id, TrainingPlanUpdateDTO data) {
        TrainingPlanModel plan = getTrainingPlan(id);
        List<ExerciseModel> exercises = exerciseService.getExercisesForUser(data.exerciseIds());

        exercises.sort(Comparator.comparingInt(
                e -> data.exerciseIds().indexOf(e.getId())
        ));

        plan.setName(data.name());
        plan.setDescription(data.description());
        plan.setExercises(exercises);

        TrainingPlanModel saved = trainingPlanRepository.save(plan);

        redisCacheService.evictUserCache(CACHE_LIST);

        return saved;
    }

    @Transactional
    public void deleteTrainingPlan(Long id) {
        UserModel user = userService.getLoggedInUser();

        TrainingPlanModel plan = trainingPlanRepository.findByIdAndUser(id, user)
                .orElseThrow(() -> new ResourceNotFound("Training plan with ID '" + id + "' does not belong to the current user"));

        trainingPlanRepository.delete(plan);
        redisCacheService.evictUserCache(CACHE_LIST);
    }
}

File: frontend/src/services/training/training-plan/training-plan.service.ts
import { api } from '../../api';
import type { PageResponseDTO } from '../../api.dto';
import type { TrainingPlanDetailsDTO, TrainingPlanDTO, TrainingPlanResponseDTO, TrainingPlanUpdateDTO } from './training-plan.dto';

const BASE_URL = '/training-plans';

export const trainingPlanService = {

	getTrainingPlans: async (page: number, size: number): Promise<PageResponseDTO<TrainingPlanResponseDTO>> => {
		try {
			const { data } = await api.get<PageResponseDTO<TrainingPlanResponseDTO>>(`${BASE_URL}?page=${page}&size=${size}`);
			return data;
		} catch (error) {
			console.error('Failed to fetch training plans.', error);
			throw new Error('Unable to retrieve training plans. Please try again.');
		}
	},

	getTrainingPlan: async (id: number): Promise<TrainingPlanDetailsDTO> => {
		try {
			const { data } = await api.get<TrainingPlanDetailsDTO>(`${BASE_URL}/${id}`);
			return data;
		} catch (error) {
			console.error(`Failed to fetch training plan with id: ${id}`, error);
			throw new Error(`Unable to retrieve training plan with id: ${id}. Please try again.`);
		}
	},

	createTrainingPlan: async (payload: TrainingPlanDTO): Promise<TrainingPlanResponseDTO> => {
		try {
			const { data } = await api.post(BASE_URL, payload);
			return data;
		} catch (error) {
			console.error('Failed to create training plan', error);
			throw new Error('Unable to create training plan. Please try again.');
		}
	},

	updateTrainingPlan: async (
		id: number,
		payload: TrainingPlanUpdateDTO
	): Promise<TrainingPlanDetailsDTO> => {
		try {
			const { data } = await api.put(`${BASE_URL}/${id}`, payload);
			return data;
		} catch (error) {
			console.error(`Failed to update training plan with id: ${id}`, error);
			throw new Error(`Unable to update training plan with id: ${id}. Please try again.`);
		}
	},

	deleteTrainingPlan: async (id: number): Promise<void> => {
		try {
			const { data } = await api.delete(`${BASE_URL}/${id}`);
			return data;
		} catch (error) {
			console.error(`Failed to delete training plan with id: ${id}`, error);
			throw new Error(`Unable to delete training plan with id: ${id}. Please try again.`);
		}
	},

}